### PR TITLE
[call creds] refactor token-fetching call creds to allow cleaner reuse

### PR DESCRIPTION
**Pull request by markdroth**
Previously, `grpc_oauth2_token_fetcher_credentials` provided functionality for on-demand token-fetching, but it was integrated into the oauth2 code, so it was not possible to use that same code for on-demand fetching of (e.g.) JWT tokens.  This PR splits that class into two parts:

1. A base `TokenFetcherCredentials` class that provides a framework for on-demand fetching of any arbitrary type of auth token.
2. An `Oauth2TokenFetcherCredentials` subclass that derives from `TokenFetcherCredentials` and provides handling for oauth2 tokens.

The `grpc_compute_engine_token_fetcher_credentials`, `StsTokenFetcherCredentials`, and `grpc_google_refresh_token_credentials` classes that previously derived from `grpc_oauth2_token_fetcher_credentials` now derive from `Oauth2TokenFetcherCredentials` instead, so there's not much change to those classes (other than a cleaner interface with the base class functionality).

The `ExternalAccountCredentials` class and its subclasses got more extensive changes here.  Previously, this class inheritted from `grpc_oauth2_token_fetcher_credentials` and fooled the base class into thinking that it directly fetched the oauth2 token, when in fact it actually performed a number of steps to gather data and then constructed a synthetic HTTP response to pass back to the base class.  I have changed this to instead derive directly from `TokenFetcherCredentials` to provide a much cleaner interface with the parent class.

In addition, I have changed `grpc_call_credentials` from `RefCounted<>` to `DualRefCounted<>` to provide a clean way to shut down any in-flight token fetch when the credentials are unreffed.

This PR paves the way for subsequent work that will allow implementing an on-demand JWT token fetcher call credential, as part of gRFC A83 (https://github.com/grpc/proposal/pull/438).

### CODE REVIEW

**Comment by markdroth** [comment]
@gtcooke94 @erm-g I've added the two of you as reviewers, since it looks like Matt is OOO.  I probably only need a review from one of you, so please feel free to decide between you who's going to do it.

I'm still tracking down a test failure on MacOS related to FuzzingEventEngine shutdown, but the overall structure of the code here shouldn't change.

File: test/core/security/credentials_test.cc [hunk 1]
```diff
@@ -2481,15 +2510,37 @@ TEST(CredentialsTest, TestExternalAccountCredsMetricsHeaderWithConfigLifetime) {
...
-
   EXPECT_EQ(
       creds.GetMetricsValue(),
       absl::StrFormat("gl-cpp/unknown auth/%s google-byoid-sdk source/unknown "
                       "sa-impersonation/true config-lifetime/true",
                       grpc_version_string()));
 }
 
-TEST(CredentialsTest, TestExternalAccountCredsSuccess) {
+using grpc_event_engine::experimental::FuzzingEventEngine;
```
Comment (matthewstevenson88) [inline test/core/security/credentials_test.cc:2520]: nit: Should we group this with the other using statements at the top of the file?

**Review by matthewstevenson88 — APPROVED** [review]
Thanks Mark!

File: src/core/lib/security/credentials/token_fetcher/token_fetcher_credentials.cc [hunk 2]
```diff
@@ -0,0 +1,130 @@
+//
+//
+// Copyright 2015 gRPC authors.
```
[the remark this answers is not in the document]
Comment (markdroth) [reply src/core/lib/security/credentials/token_fetcher/token_fetcher_credentials.cc:3]: I basically branched this from the existing code, so I retained the original copyright date.

File: src/core/lib/security/credentials/token_fetcher/token_fetcher_credentials.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
+//
+// Copyright 2016 gRPC authors.
```
[the remark this answers is not in the document]
Comment (markdroth) [reply src/core/lib/security/credentials/token_fetcher/token_fetcher_credentials.h:2]: Same here.

File: test/core/security/credentials_test.cc (hunk 1 above)
Comment (markdroth) [reply test/core/security/credentials_test.cc:2520]: I'll take care of that in a subsequent PR.

**Pull request closed without merging**